### ISSUE TITLE
Extending Storage Access API (SAA) to non-cookie storage

### ISSUE THREAD

**Issue by arichiv** [mozilla/standards-positions#898]
## Request for Mozilla Position on an Emerging Web Specification

* Specification title: Extending Storage Access API (SAA) to non-cookie storage
* Specification or proposal URL (if available): https://privacycg.github.io/saa-non-cookie-storage/
* Proposal author(s) (`@`-mention GitHub accounts): @arichiv @johannhof 
* Bugtracker URL: https://crbug.com/1484966
* Safari Position: https://github.com/WebKit/standards-positions/issues/262
* TAG:  https://github.com/w3ctag/design-reviews/issues/906

### Other information

We propose an extension of the Storage Access API (backwards compatible) to allow access to unpartitioned (cookie and non-cookie) storage in a third-party context, and imagine the API mechanics to be roughly like this (JS running in an embedded iframe):

// Request a new storage handle via rSA (this should prompt the user)
let handle = await document.requestStorageAccess({all: true});
// Write some cross-site localstorage
handle.localStorage.setItem("userid", "1234");
// Open or create an indexedDB that is shared with the 1P context
let messageDB = handle.defaultBucket.indexedDB.open("messages");

The same flow would be used by iframes to get a storage handle when their top-level ancestor successfully called rSAFor, just that in this case the storage-access permission was already granted and thus the rSA call would not require a user gesture or show a prompt, allowing for “hidden” iframes accessing storage.

**Comment by zcorpan** [comment]
@asutherland

**Comment by asutherland** [comment]
@bvandersloot-mozilla (StorageAccess is primarily a privacy/anti-tracking concern).

**Comment by bvandersloot-mozilla** [comment]
See the [original discussion](https://github.com/privacycg/storage-access/issues/102) and [Privacy CG proposal](https://github.com/privacycg/proposals/issues/41).

There is no privacy concern here. Unpartitioned storage offers no benefit to attackers over unpartitioned cookies. Instead the question is whether the API is reasonable. @asutherland and @annevk offered varying degrees of pushback on localStorage and sessionStorage because they are sync. 

I believe this proposal resolves @asutherland's concerns though. This call returns a handle, rather than trying to use storage buckets. Each storage type is opt-in, making it a developer choice if it is acceptable to degrade performance. But, it may be worth a second look.

**Comment by asutherland** [comment]
I'm generally happy with the consensus direction we ended up with in https://github.com/privacycg/storage-access/issues/102 through the current timestamp.

The API described in the [proposed solution](https://arichiv.github.io/saa-non-cookie-storage/#proposed-solution) seems consistent-ish with that but I would like to see the proposal expanded on, especially in terms of the semantics of "all" which was not/has not yet been discussed on https://github.com/privacycg/storage-access/issues/102.  Specifically, I can see an advantage to developers of being able to say "I want all the new storage API endpoints that show up in the future too", but it would be nice then to be sure that something like `{ all: true, localStorage: false }` works so that developers can get the benefit of "all" but without incurring the cost of LocalStorage.

I know it's frequently explicitly discouraged for people to put WebIDL in explainers, but since Blink seems to be at the prototyping stage already, it likely does make sense to have some draft WebIDL and spec language around the relationship to the Storage spec's endpoint terminology.

**Comment by arichiv** [comment]
I agree that 'all' has tenuous value, to some extent it's there for developers who are testing or prefer convenience. I'm not sure I like the idea of an 'all, except for' feature. Developers who aren't into the convenience of 'all' should probably just be requesting the specific APIs they want.

I agree about WebIDL, if/once the explainer is adopted by the PrivacyCG I plan to put up a PR to modify the existing spec: https://privacycg.github.io/storage-access/.

**Comment by arichiv** [comment]
Two additional explainers (each of which is an extension to [Storage Access API (SAA) to non-cookie storage](https://arichiv.github.io/saa-non-cookie-storage/)) have been published!

[Explainer: Extending Storage Access API (SAA) to omit unpartitioned cookies](https://arichiv.github.io/saa-non-cookie-storage/omit-unpartitioned-cookies.html)
The current Storage Access API requires that unpartitioned cookie access is granted if any unpartitioned storage access is needed. This forces unpartitioned cookies to be included in network requests which may not need them, having impacts on network performance and security. Before the extension ships, we have a chance to fix this behavior without a compatibility break.

[Explainer: Extending Storage Access API (SAA) to Shared Workers](https://arichiv.github.io/saa-non-cookie-storage/shared-workers.html)
There has been increasing developer and implementer interest in first-party workers being available in third-party contexts the same way that third-party cookies already can be. In the absence of such a solution, we leave developers without a robust way to manage cross-tab state for frames loading the same origin. This explainer proposes a solution for developers to regain third-party access to Shared Workers in select instances to avoid user-facing breakage in browsers shipping storage partitioning.

Would it be possible to get review on this issue of those as well?

**Comment by bvandersloot-mozilla** [comment]
Circling back through unlabeled issues... I think we landed somewhere positive for this, as much as we can with the current level of detail. Having such an escape hatch to prevent over-use of cookies seems good where there is no privacy property change to using Storage.

**Comment by sunetzacharias** [comment]
We are rolling this out in production in the coming days, and it seems to be only within Chrome this is fully supported. As such we are reaching out to you and Safari.

Is there a timeline for this?

We are SeamlessAccess, a discovery service supporting research and higher education which covers metadata from an interfederation consisting of more than 5000 IdPs.

We offer a persistence service to store IdP metadata in a way that can be shared among Service Providers SPs, so that a choice of IdP to login to some SP can be reused to login to a different SP without needing to go through the discovery process again. This is implemented as an iframe with origin in a SeamlessAccess domain that interacts with the embedding SP pages through message events, and stores the metadata in localStorage.
With partitioned storage we will be forced to use the SAA to access non-partitioned storage and provide the persistence service, but the amount of data we need to store will typically not fit in a cookie.

You can see it in action with the Storage Access API tooling here [use.thiss.io](use.thiss.io)
